types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+		struct wlr_drm_lease_connector_v1 *connector,
+		struct wl_client *wl_client, struct wl_resource *device) {
+	if (connector->active_lease) {
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(wl_client,
+			&wp_drm_lease_connector_v1_interface, DRM_LEASE_DEVICE_V1_VERSION,
+			0);
+	if (!wl_resource) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(wl_resource, &lease_connector_impl,
+			connector, drm_connector_v1_handle_resource_destroy);
+	wp_drm_lease_device_v1_send_connector(device, wl_resource);
+
+	struct wlr_output *output = connector->output;
+	wp_drm_lease_connector_v1_send_name(wl_resource, output->name);
+
+	char description[128];
+	snprintf(description, sizeof(description), "%s %s %s",
+		output->make, output->model, output->serial);
+	wp_drm_lease_connector_v1_send_description(wl_resource, description);
+
+	struct wlr_drm_connector *drm_connector =
+			get_drm_connector_from_output(output);
+	wp_drm_lease_connector_v1_send_connector_id(wl_resource, drm_connector->id);
+
+	wl_list_insert(&connector->resources, wl_resource_get_link(wl_resource));
+}
+
+static void lease_device_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device = data;
+
+	struct wl_resource *wl_resource  = wl_resource_create(wl_client,
+		&wp_drm_lease_device_v1_interface, version, id);
+
+	if (!wl_resource) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_resource_set_implementation(wl_resource, &lease_device_impl, device,
+			drm_lease_device_v1_handle_resource_destroy);
+
+	int fd = wlr_drm_backend_get_read_only_fd(device->backend);
+	if (fd < 0) {
+		wlr_log(WLR_ERROR, "Unable to get read only DRM fd for leasing");
+		return;
+	}
+
+	wp_drm_lease_device_v1_send_drm_fd(wl_resource, fd);
+	close(fd);
+
+	wl_list_insert(&device->resources, wl_resource_get_link(wl_resource));
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		drm_lease_connector_v1_send_to_client(connector, wl_client,
+				wl_resource);
+	}
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_connector_v1 *conn = wl_container_of(listener, conn,
+			destroy);
+
+	wlr_log(WLR_DEBUG, "Destroying connector %s", conn->output->name);
+	if (conn->active_lease) {
+		wlr_drm_lease_v1_revoke(conn->active_lease);
+	}
+	wlr_drm_lease_manager_withdraw_output(conn->device->manager, conn->output);
+}
+
+bool wlr_drm_lease_manager_offer_output(
+		struct wlr_drm_lease_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+	assert(wlr_output_is_drm(output));
+
+	wlr_log(WLR_DEBUG, "Offering output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL;
+	wl_list_for_each(device, &manager->devices, link) {
+		if (device->backend == output->backend) {
+			break;
+		}
+	}
+
+	if (!device) {
```

Comment:
This check will never catch the error case. If the device isn't found, `device` will be set to the last device after the loop stops.